tools/patch.py

```diff
@@ -0,0 +1,208 @@
+# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+# PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+# OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+"""
+This is a basic patch tool that can parse unified diff format patches
+and apply them to files. It is not a full-featured patch tool and
+only supports a subset of the unified diff format, just enough to
+apply the patches we need. It won't properly validate the patch file
+format, so it's up to the user to ensure the patch file is correct.
+It's also not optimized for performance, so it may be slow for large
+patches or files.
+"""
+
+import os
+import sys
+from enum import Enum
+
+class FileAlreadyPatchedError(Exception):
+    pass
+
+class FileNotMatchingError(Exception):
+    pass
+
+class PatchMalformedError(Exception):
+    pass
+
+class LineType(Enum):
+    MATCH = ' '
+    ADD = '+'
+    REMOVE = '-'
+
+class PatchLine:
+    def __init__(self, line, line_type):
+        self.line = line
+        self.line_type = line_type
+
+    def __str__(self):
+        return f"{self.line_type.value} {self.line}"
+
+class Hunk:
+    def __init__(self, filename, location, count):
+        self.filename = filename
+        self.location = int(location)
+        self.count = int(count)
+        self.lines = []
+
+    def apply(self):
+        with open(self.filename, 'r') as file:
+            lines = file.readlines()
+
+        output = []
+        current_line = 1
+        relative_line = 0
+        for line in lines:
+            line = line.rstrip('\r\n')
+            # Pass through the lines not inside the hunk
+            if current_line < int(self.location) or current_line >= int(self.location + self.count):
+                output.append(line)
+            else:
+                if self.lines[relative_line].line_type == LineType.MATCH:
+                    if self.lines[relative_line].line != line:
+                        raise FileNotMatchingError(f"Patch failed: {self.filename}:{current_line} does not match expected line:\nSource: {line}\nTarget: {self.lines[relative_line].line}")
+                    output.append(line)
+                    relative_line += 1
+                elif self.lines[relative_line].line_type == LineType.ADD:
+                    # Verify we're not already patched
+                    if self.lines[relative_line].line == line:
+                        raise FileAlreadyPatchedError(f"Patch failed: {self.filename}:{current_line} already patched")
+                    # Write lines as long as we have a streak of ADD lines
+                    while self.lines[relative_line].line_type == LineType.ADD:
```

Comment:
Can't this loop be applied to all line types? Is there a particular reason we only apply it to ADD?

Reply:
So, there is a discrepancy between "input lines" and "patch lines" here in this context. There is only a single "input line", and there might be many corresponding patch-ADD lines.

It makes more sense when looking at it in action: for https://github.com/triton-inference-server/third_party/blob/main/tools/abseil_no_rebuild.patch#L16-L27, the input lines will be

```
  endif()
  file(WRITE "${CMAKE_BINARY_DIR}/options-pinned.h" "${ABSL_INTERNAL_OPTIONS_H_PINNED}")

```
(followed by an empty line).

So when we're at the line `file(WRITE...`, the corresponding patch information tells us to remove it, so we skip it, and then when we're at the empty line right after, the corresponding patch information tells us to add a bunch of lines prior to it, hence the while loop.

The "REMOVE" and "MATCH" commands will always be 1:1 from the input lines, that is. "ADD" is the only one which has no correspondence, and that we have to fully consume before moving on to the next.

Reply:
Ah I see. Thank you for the thorough explanation!